formats/json/commonMain/src/kotlinx/serialization/json/internal/TreeJsonDecoder.kt

```diff
@@ -224,18 +224,21 @@ private open class JsonTreeDecoder(
         return !forceNull && super.decodeNotNullMark()
     }
 
-    override fun elementName(desc: SerialDescriptor, index: Int): String {
-        val mainName = desc.getElementName(index)
-        if (!configuration.useAlternativeNames) return mainName
-        // Fast path, do not go through ConcurrentHashMap.get
-        // Note, it blocks ability to detect collisions between the primary name and alternate,
-        // but it eliminates a significant performance penalty (about -15% without this optimization)
-        if (mainName in value.keys) return mainName
+    override fun elementName(descriptor: SerialDescriptor, index: Int): String {
+        val strategy = descriptor.namingStrategy(json)
+        val baseName = descriptor.getElementName(index)
+        if (strategy == null) {
+            if (!configuration.useAlternativeNames) return baseName
+            // Fast path, do not go through ConcurrentHashMap.get
+            // Note, it blocks ability to detect collisions between the primary name and alternate,
+            // but it eliminates a significant performance penalty (about -15% without this optimization)
+            if (baseName in value.keys) return baseName
+        }
         // Slow path
-        val alternativeNamesMap =
-            json.schemaCache.getOrPut(desc, JsonAlternativeNamesKey, desc::buildAlternativeNamesMap)
-        val nameInObject = value.keys.find { alternativeNamesMap[it] == index }
-        return nameInObject ?: mainName
+        val deserializationNamesMap = json.deserializationNamesMap(descriptor)
+        val nameInObject = value.keys.find { deserializationNamesMap[it] == index }
+        val fallbackName = strategy?.serialNameForJson(descriptor, index, descriptor.getElementName(index)) // Key not found exception should be thrown with transformed named, not original
```

Comment:
Useless invoke if `nameInObject != null`, so it can be optimized `nameInObject?.let { return it }`